File: 1025-minimum-cost-for-tickets/1025-minimum-cost-for-tickets.cs
public class Solution {
    public int solve(int i,int[] days,int[] costs,int[] dp){
        if(i >= days.Length){
            return 0;
        }
        if(dp[i] != -1)return dp[i];
        int value = days[i];
        int indxForOneDay = Array.BinarySearch(days,value+1);
        int indxForSevenDay = Array.BinarySearch(days,value+7);
        int indxForThirtyDay = Array.BinarySearch(days,value+30);

        if(indxForOneDay<0) indxForOneDay = ~indxForOneDay;
        if(indxForSevenDay<0) indxForSevenDay = ~ indxForSevenDay;
        if(indxForThirtyDay<0) indxForThirtyDay = ~ indxForThirtyDay;

        int ans = 1000000000 + 7;
        ans = Math.Min(ans,costs[0] + solve(indxForOneDay,days,costs,dp));
        ans = Math.Min(ans,costs[1] + solve(indxForSevenDay,days,costs,dp));
        ans = Math.Min(ans,costs[2] + solve(indxForThirtyDay,days,costs,dp));
        return dp[i] = ans;
    }
    public int MincostTickets(int[] days, int[] costs) {
        int []dp = new int[370];
        Array.Fill(dp,-1);
        return solve(0,days,costs,dp);
    }
}

File: 1025-minimum-cost-for-tickets/1025-minimum-cost-for-tickets.cpp
class Solution {
public:
    int solve(int i , vector<int>&days, vector<int>& costs,vector<int>&dp){
        if(i>= days.size())return 0;
        int value = days[i];
        if(dp[i] !=-1 )return dp[i];
        int indx0 = lower_bound(days.begin(),days.end(),value+1)-days.begin();
        int indx = lower_bound(days.begin(),days.end(),value+7)-days.begin();
        int indx1 = lower_bound(days.begin(),days.end(),value+30)-days.begin();
        int ans = 1e9;
        ans = min(ans,costs[0] + solve(indx0,days,costs,dp));
        ans = min(ans , costs[1] + solve(indx,days,costs,dp));
        ans = min(ans , costs[2] + solve(indx1 ,days,costs,dp));
        return dp[i] = ans;
    }
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        //vector<vector<int>>dp(days.size()+1,vecctor<int>(,))
        vector<int>dp(367,-1);
        return solve(0,days,costs,dp);
    }
};